pkg/providers/cloudstack/cloudstack.go

```diff
@@ -253,69 +253,49 @@ func (p *cloudstackProvider) MachineResourceType() string {
 	return eksaCloudStackMachineResourceType
 }
 
-func (p *cloudstackProvider) setupSSHAuthKeysForCreate() error {
-	var useKeyGeneratedForControlplane, useKeyGeneratedForWorker bool
-	var err error
-	controlPlaneUser := p.machineConfigs[p.clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec.Users[0]
-	if len(controlPlaneUser.SshAuthorizedKeys[0]) > 0 {
-		controlPlaneUser.SshAuthorizedKeys[0], err = common.StripSshAuthorizedKeyComment(controlPlaneUser.SshAuthorizedKeys[0])
+func (p *cloudstackProvider) setupSSHAuthKeysForCreateOrUpgrade() error {
+	var generatedKey string
+	generatedKey, err := p.processUsers(p.machineConfigs[p.clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec.Users, generatedKey)
+	if err != nil {
+		return err
+	}
+	for _, workerNodeGroupConfiguration := range p.clusterConfig.Spec.WorkerNodeGroupConfigurations {
+		generatedKey, err = p.processUsers(p.machineConfigs[workerNodeGroupConfiguration.MachineGroupRef.Name].Spec.Users, generatedKey)
 		if err != nil {
 			return err
 		}
-	} else {
-		logger.Info("Provided control plane sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
-		generatedKey, err := common.GenerateSSHAuthKey(p.writer)
+	}
+	if p.clusterConfig.Spec.ExternalEtcdConfiguration != nil {
+		_, err = p.processUsers(p.machineConfigs[p.clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users, generatedKey)
 		if err != nil {
 			return err
 		}
-		controlPlaneUser.SshAuthorizedKeys[0] = generatedKey
-		useKeyGeneratedForControlplane = true
 	}
-	var workerUser v1alpha1.UserConfiguration
-	for _, workerNodeGroupConfiguration := range p.clusterConfig.Spec.WorkerNodeGroupConfigurations {
-		workerUser = p.machineConfigs[workerNodeGroupConfiguration.MachineGroupRef.Name].Spec.Users[0]
-		if len(workerUser.SshAuthorizedKeys[0]) > 0 {
-			workerUser.SshAuthorizedKeys[0], err = common.StripSshAuthorizedKeyComment(workerUser.SshAuthorizedKeys[0])
-			if err != nil {
-				return err
-			}
-		} else {
-			if useKeyGeneratedForControlplane { // use the same key
-				workerUser.SshAuthorizedKeys[0] = controlPlaneUser.SshAuthorizedKeys[0]
-			} else {
-				logger.Info("Provided worker sshAuthorizedKey is not set or is empty, auto-generating new key pair...")
-				generatedKey, err := common.GenerateSSHAuthKey(p.writer)
+	return nil
+}
+
+func (p *cloudstackProvider) processUsers(users []v1alpha1.UserConfiguration, generatedKey string) (string, error) {
```

Comment:
could we rename to something like `processSshKeysForUser` instead? First, because it only processes a single user, and second so that we can include how we're processing the user

Reply:
I just renamed it to processSshKeysForUsers because it processes multiple users. `users []v1alpha1.UserConfiguration`